src/mountinfo.c

```diff
@@ -0,0 +1,276 @@
+}
+
+const char *mountinfo_entry_optional_fields(struct mountinfo_entry *entry)
+{
+	return entry->optional_fields;
+}
+
+const char *mountinfo_entry_fs_type(struct mountinfo_entry *entry)
+{
+	return entry->fs_type;
+}
+
+const char *mountinfo_entry_mount_source(struct mountinfo_entry *entry)
+{
+	return entry->mount_source;
+}
+
+const char *mountinfo_entry_super_opts(struct mountinfo_entry *entry)
+{
+	return entry->super_opts;
+}
+
+struct mountinfo *parse_mountinfo(const char *fname)
+{
+	if (fname == NULL) {
+		fname = "/proc/self/mountinfo";
+	}
+	FILE *f __attribute__ ((cleanup(cleanup_fclose))) = fopen(fname, "rt");
+	if (f == NULL) {
+		return NULL;
+	}
+	struct mountinfo *info = calloc(1, sizeof *info);
+	if (info == NULL) {
+		return NULL;
+	}
+	char *line __attribute__ ((cleanup(cleanup_free))) = NULL;
+	size_t line_size = 0;
+	struct mountinfo_entry *entry, *last = NULL;
+	for (;;) {
+		errno = 0;
+		if (getline(&line, &line_size, f) == -1) {
+			if (errno != 0) {
+				free_mountinfo(info);
+				return NULL;
+			}
+			break;
+		};
+		entry = parse_mountinfo_entry(line);
+		if (entry == NULL) {
+			free_mountinfo(info);
+			return NULL;
+		}
+		if (last != NULL) {
+			last->next = entry;
+		} else {
+			info->first = entry;
+		}
+		last = entry;
+	}
+	return info;
+}
+
+static struct mountinfo_entry *parse_mountinfo_entry(const char *line)
+{
+	struct mountinfo_entry *entry =
+	    calloc(1, sizeof *entry + strlen(line) + 1);
+	if (entry == NULL) {
+		return NULL;
+	}
+	int nscanned;
+	int offset, total_offset = 0;
+	nscanned = sscanf(line, "%d %d %u:%u %n",
+			  &entry->mount_id, &entry->parent_id,
+			  &entry->dev_major, &entry->dev_minor, &offset);
+	if (nscanned != 4)
+		goto fail;
+	total_offset += offset;
+	int total_used = 0;
+	char *parse_next_string_field() {
+		char *field = &entry->line_buf[0] + total_used;
+		nscanned = sscanf(line + total_offset, "%s %n", field, &offset);
+		if (nscanned != 1)
+			return NULL;
+		total_offset += offset;
+		total_used += offset + 1;
+		return field;
+	}
+	if ((entry->root = parse_next_string_field()) == NULL)
+		goto fail;
+	if ((entry->mount_dir = parse_next_string_field()) == NULL)
+		goto fail;
+	if ((entry->mount_opts = parse_next_string_field()) == NULL)
+		goto fail;
+	entry->optional_fields = &entry->line_buf[0] + total_used++;
+	strcpy(entry->optional_fields, "");
+	for (;;) {
+		char *opt_field = parse_next_string_field();
+		if (opt_field == NULL)
+			goto fail;
+		if (strcmp(opt_field, "-") == 0) {
+			break;
+		}
+		if (*entry->optional_fields) {
+			strcat(entry->optional_fields, " ");
+		}
+		strcat(entry->optional_fields, opt_field);
+	}
+	if ((entry->fs_type = parse_next_string_field()) == NULL)
+		goto fail;
+	if ((entry->mount_source = parse_next_string_field()) == NULL)
+		goto fail;
+	if ((entry->super_opts = parse_next_string_field()) == NULL)
+		goto fail;
+	return entry;
+ fail:
+	free(entry);
+	return NULL;
+}
+
+void cleanup_mountinfo(struct mountinfo **ptr)
```

Comment:
This is a library function (for users of mount info), note that free_mountinfo is static/private.